### PR TITLE
Fix multi_hot and unfold in graph mode for OpenVINO and TensorFlow

### PR DESCRIPTION
**Pull request by goyaladitya05**
## Description
`ops.multi_hot` and `ops.unfold` failed under symbolic graph tracing, i.e. any functional model or model.predict, since keras.Input always has a dynamic batch dim. Only eager calls with fully static shapes worked, which is what the existing tests do, so CI never caught it.

- OpenVINO multi_hot/unfold passed an OpenVINOKerasTensor straight into ov_opset, which falls back to np.array() -> __array__ -> convert_to_numpy. On constants that silently compiled a model at graph-build time and froze the result into a Constant, on a Parameter it raised.

- OpenVINO and TensorFlow unfold used the batch dim read off input.shape directly in reshape, which is None/dynamic during tracing. Replaced with -1. numpy/jax/torch are unaffected because they trace with concrete batch sizes.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/backend/openvino/nn.py [hunk 1]
```diff
@@ -1352,27 +1354,35 @@ def _pair(x):
...
+    # `-1` for the batch dim keeps the graph valid when it is dynamic, as it
+    # is under model.predict.
+    patches = ov_opset.reshape(
+        patches, [-1, k[0], k[1], C, nH, nW], False
+    ).output(0)
+    # (N, C, kH, kW, nH, nW)
+    patches = ov_opset.transpose(patches, [0, 3, 1, 2, 4, 5]).output(0)
+    patches = ov_opset.reshape(
+        patches, [-1, C * k[0] * k[1], nH * nW], False
+    ).output(0)
```
[the remark this answers is not in the document]
Comment (goyaladitya05) [reply keras/src/backend/openvino/nn.py:1384]: the path isn't reachable today

File: keras/src/backend/tensorflow/nn.py [hunk 2]
```diff
@@ -1886,13 +1886,15 @@ def unfold(input, kernel_size, dilation=1, padding=0, stride=1):
...
+    # is for any functional model or model.predict.
     patches = tf.reshape(
-        patches, [N, nH, nW, k[0], k[1], C]
+        patches, [-1, nH, nW, k[0], k[1], C]
     )  # (N, nH, nW, kH, kW, C)
     patches = tf.transpose(
         patches, [0, 5, 3, 4, 1, 2]
     )  # (N, C, kH, kW, nH, nW)
-    patches = tf.reshape(patches, [N, C * k[0] * k[1], nH * nW])
+    patches = tf.reshape(patches, [-1, C * k[0] * k[1], nH * nW])
```
[the remark this answers is not in the document]
Comment (goyaladitya05) [reply keras/src/backend/tensorflow/nn.py:1897]: the path isn't reachable today

File: keras/src/ops/nn_test.py [hunk 3]
```diff
@@ -313,6 +313,44 @@ def test_multi_hot(self):
...
+    def test_unfold_in_graph(self):
+        class UnfoldLayer(keras.Layer):
+            def call(self, x):
+                return ops.unfold(x, kernel_size=2)
+
+        class Model(keras.Model):
+            def __init__(self):
+                x = keras.Input(shape=(2, 6, 6))
+                y = UnfoldLayer()(x)
+                super().__init__(inputs=x, outputs=y)
```
[the remark this answers is not in the document]
Comment (goyaladitya05) [reply keras/src/ops/nn_test.py:347]: this dynamic spatial dims are untested, because they're unsupported.

File: keras/src/ops/nn_test.py [hunk 4]
```diff
@@ -313,6 +313,44 @@ def test_multi_hot(self):
...
+                return (input_shape[0], 5)
+
+        class Model(keras.Model):
+            def __init__(self):
+                x = keras.Input(shape=(3,), dtype="int32")
+                y = MultiHotLayer()(x)
+                super().__init__(inputs=x, outputs=y)
+
+        # Make sure Keras is able to compile the model graph
+        model = Model()
```
Comment (hertschuh) [inline keras/src/ops/nn_test.py:334]: You should just be able to do this instead of all this code:
```python
x = keras.Input(shape=(3,), dtype="int32")
y = knn.multi_hot(x, 5)
model = Model(inputs=x, outputs=y)
```

File: keras/src/ops/nn_test.py [hunk 5]
```diff
@@ -313,6 +313,44 @@ def test_multi_hot(self):
...
+                return ops.unfold(x, kernel_size=2)
+
+        class Model(keras.Model):
+            def __init__(self):
+                x = keras.Input(shape=(2, 6, 6))
+                y = UnfoldLayer()(x)
+                super().__init__(inputs=x, outputs=y)
+
+        # Make sure Keras is able to compile the model graph
+        model = Model()
```
Comment (hertschuh) [inline keras/src/ops/nn_test.py:350]: You should just be able to do this instead of all this code:
```python
x = keras.Input(shape=(3,), dtype="int32")
y = knn.unfold(x, kernel_size=2)
model = Model(inputs=x, outputs=y)
```

File: keras/src/ops/nn_test.py [hunk 6]
```diff
@@ -313,6 +313,44 @@ def test_multi_hot(self):
...
+        class Model(keras.Model):
+            def __init__(self):
+                x = keras.Input(shape=(3,), dtype="int32")
+                y = MultiHotLayer()(x)
+                super().__init__(inputs=x, outputs=y)
+
+        # Make sure Keras is able to compile the model graph
+        model = Model()
+        x = ops.array([[0, 2, 4]], dtype="int32")
+        model.predict(x)
```
Comment (hertschuh) [inline keras/src/ops/nn_test.py:336]: I would add:
```python
        model.run_eagerly = False
        model.jit_compile = True
```
Because I think it won't compile in some combinations of cases (e.g. TensorFlow on CPU).

File: keras/src/ops/nn_test.py [hunk 7]
```diff
@@ -313,6 +313,44 @@ def test_multi_hot(self):
...
+        class Model(keras.Model):
+            def __init__(self):
+                x = keras.Input(shape=(2, 6, 6))
+                y = UnfoldLayer()(x)
+                super().__init__(inputs=x, outputs=y)
+
+        # Make sure Keras is able to compile the model graph
+        model = Model()
+        x = ops.ones((1, 2, 6, 6))
+        model.predict(x)
```
Comment (hertschuh) [inline keras/src/ops/nn_test.py:352]: I would add:
```python
        model.run_eagerly = False
        model.jit_compile = True
```
Because I think it won't compile in some combinations of cases (e.g. TensorFlow on CPU).

File: keras/src/backend/tensorflow/nn.py (hunk 2 above)
Comment (hertschuh) [inline keras/src/backend/tensorflow/nn.py:1897]: Actually, here is another way to fix this that I like a bit better:

Add this at the top:
```python
from keras.src.backend.tensorflow.core import shape as shape_op
```

Then change this:
```python
N, nH, nW, D = shape_op(patches)
```

Then change back the `-1` to `N`.

`shape_op` extract tensors for dynamic dimensions, so `N` will either be an int or a tensor.

The reason I like it better is that it will work also if there is more dimensions than the batch size that's dynamic, where the `-1` trick assumes the batch size is the only dynamic dimension.

**Review by hertschuh — CHANGES_REQUESTED** [review]
